### PR TITLE
REGR: overlay keep_geom_type issue with collections of different types

### LINKED ISSUE #2176: BUG: 0.10 overlay fails where 0.9 succeeded
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of geopandas.

- [ ] (optional) I have confirmed this bug exists on the master branch of geopandas.

---

#### Code Sample, a copy-pastable example

Data gzipped so github would allow upload: [overlay_bug.gpkg.gz](https://github.com/geopandas/geopandas/files/7337850/overlay_bug.gpkg.gz)

```python
import geopandas as gpd

gdf = gpd.read_file("overlay_bug.gpkg", driver="GPKG")
overlay_gdf = gdf.overlay(gdf, how="intersection")
```

Gives
```
Traceback (most recent call last):
  File "bug_report_2.py", line 4, in <module>
    overlay_gdf = gdf.overlay(gdf, how="intersection")
  File "/opt/conda/envs/pyai/lib/python3.8/site-packages/geopandas/geodataframe.py", line 2196, in overlay
    return geopandas.overlay(
  File "/opt/conda/envs/pyai/lib/python3.8/site-packages/geopandas/tools/overlay.py", line 358, in overlay
    result.loc[is_collection, geom_col] = dissolved[geom_col].values
  File "/opt/conda/envs/pyai/lib/python3.8/site-packages/pandas/core/indexing.py", line 723, in __setitem__
    iloc._setitem_with_indexer(indexer, value, self.name)
  File "/opt/conda/envs/pyai/lib/python3.8/site-packages/pandas/core/indexing.py", line 1730, in _setitem_with_indexer
    self._setitem_with_indexer_split_path(indexer, value, name)
  File "/opt/conda/envs/pyai/lib/python3.8/site-packages/pandas/core/indexing.py", line 1785, in _setitem_with_indexer_split_path
    raise ValueError(
ValueError: Must have equal len keys and value when setting with an iterable
```

#### Problem description

New and unexpected failure

#### Expected Output

Overlay should complete, or at least fail gracefully with error message.

#### Output of ``geopandas.show_versions()``

<details>

SYSTEM INFO
-----------
python     : 3.8.12 | packaged by conda-forge | (default, Oct 12 2021, 21:59:51)  [GCC 9.4.0]
executable : /opt/conda/envs/pyai/bin/python
machine    : Linux-5.10.47-linuxkit-x86_64-with-glibc2.10

GEOS, GDAL, PROJ INFO
---------------------
GEOS       : 3.9.1
GEOS lib   : /opt/conda/envs/pyai/lib/libgeos_c.so
GDAL       : 3.3.1
GDAL data dir: /opt/conda/envs/pyai/share/gdal
PROJ       : 8.0.1
PROJ data dir: /opt/conda/envs/pyai/share/proj

PYTHON DEPENDENCIES
-------------------
geopandas  : 0.10.1
pandas     : 1.3.3
fiona      : 1.8.20
numpy      : 1.21.2
shapely    : 1.7.1
rtree      : 0.9.7
pyproj     : 3.2.1
matplotlib : 3.4.3
mapclassify: 2.4.3
geopy      : 2.2.0
psycopg2   : 2.9.1 (dt dec pq3 ext lo64)
geoalchemy2: None
pyarrow    : 5.0.0
pygeos     : 0.10.2

</details>

### PR DESCRIPTION
**Pull request by martinfleis**
Fixes #2176 

I still need to fabricate some tests, this is a very edge case.

### CODE REVIEW

File: geopandas/tools/overlay.py [hunk 1]
```diff
@@ -361,6 +363,7 @@ def _make_valid(df):
             # respective Multi version)
             dissolved = exploded.dissolve(by="level_0")
             result.loc[is_collection, geom_col] = dissolved[geom_col].values
+            result = result[result.geometry.notna()]
```
Comment (jorisvandenbossche) [inline geopandas/tools/overlay.py:366]: This might not be needed, because below we still do a filter operation to only keep the geometries of a certain type (for which None will always gives False, and get filtered out)
